Semantic snippets: Implement `do` snippet

DoctorKrolic opened the pull request:
Part of https://github.com/dotnet/roslyn/issues/64144 (don't forget to check the item when merging)

CyrusNajmabadi commented on src/Features/CSharp/Portable/Completion/CompletionProviders/SnippetCompletionProvider.cs:
```diff
@@ -36,6 +36,7 @@ internal sealed class SnippetCompletionProvider : LSPCompletionProvider
         "class",
         "cw",
         "ctor",
+        "do",
```
fwiw, these should likely be constants.

CyrusNajmabadi commented on src/Features/CSharp/Portable/Snippets/CSharpDoWhileLoopStatementProvider.cs:
```diff
@@ -0,0 +1,61 @@
+[method: Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+internal sealed class CSharpDoWhileLoopStatementProvider() : AbstractConditionalBlockSnippetProvider
+{
+    public override string Identifier => "do";
```
with the constant referenced here.

CyrusNajmabadi commented on src/Features/CSharp/Portable/Snippets/CSharpDoWhileLoopStatementProvider.cs:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    protected override Func<SyntaxNode?, bool> GetSnippetContainerFunction(ISyntaxFacts syntaxFacts)
+        => static (node) => node is DoStatementSyntax;
```
```suggestion
        => static node => node is DoStatementSyntax;
```

DoctorKrolic commented:
@CyrusNajmabadi Regardless your request on extracting constants: I tried to preserve layering and separated identifiers into 2 categories: those which can theoretically be reused in VB (e.g. `cw`) and those which are C# only (keywords and so on). As part of that work I moved some properties up to C# layer (e.g. for `else` identifier since it is clearly a C# keyword and in VB it would be `Else`, same with description). Also since I touched most files in snippets folder anyway, I refactored classes in it a bit by making importing constructors primary and sealing what was still unsealed

DoctorKrolic commented:
@CyrusNajmabadi PTAL

DoctorKrolic commented:
@CyrusNajmabadi PTAL

CyrusNajmabadi commented on src/Features/CSharp/Portable/Snippets/CSharpDoWhileLoopStatementProvider.cs:
```diff
@@ -0,0 +1,61 @@
+
+    protected override SyntaxNode GetCondition(SyntaxNode node)
+    {
+        var doStatement = (DoStatementSyntax)node;
```
nit: int eh future, it would nice to use generics (with teh base type) to make it so that all this is strongly typed.

DoctorKrolic replied:
In this particular example this would just move the hard cast one layer above:
https://github.com/dotnet/roslyn/blob/e6c68774586f44370083f0f444a1b165b0ef7182/src/Features/Core/Portable/Snippets/SnippetProviders/AbstractConditionalBlockSnippetProvider.cs#L21-L26
`GetPlaceHolderLocationsList` is from the `AbstractSnippetProvider` class, which is a base class for all snippets. And even if we try to make `AbstractSnippetProvider` generic as well, we will still need to hard cast here (see how `mainChangeNode` is just found via annotation, the cast would be here):
https://github.com/dotnet/roslyn/blob/e6c68774586f44370083f0f444a1b165b0ef7182/src/Features/Core/Portable/Snippets/SnippetProviders/AbstractSnippetProvider.cs#L110-L123

CyrusNajmabadi replied:
Yup.  mkaing these changes in https://github.com/dotnet/roslyn/pull/72892
